### PR TITLE
Refs #33646 -- Made QuerySet.raw() async-compatible.

### PR DESCRIPTION
**Pull request by felixxm**
ticket-33646

### CODE REVIEW

**Comment by felixxm** [comment]
The 1st commit is crucial to reusing async-logic from `BaseIterable`.

File: django/db/models/query.py [hunk 1]
```diff
@@ -1981,6 +2025,12 @@ def _fetch_all(self):
         if self._prefetch_related_lookups and not self._prefetch_done:
             self._prefetch_related_objects()
 
+    async def _async_fetch_all(self):
+        if self._result_cache is None:
+            self._result_cache = [result async for result in RawModelIterable(self)]
```
Comment (carltongibson) [inline django/db/models/query.py:2030]: Note in passing: 

(Already in play from #14843, but _interesting_, to me at least 😜) 

The `__aiter__` here calls `BaseIterable._async_generator()` ([ln50](https://github.com/django/django/pull/15637/files#diff-d58ef61559dc7af5fdf7b56fee13571a4d2948e784cd608f6afeacf3ac2fb195R50)) which has this nice step: 

```
        while True:
            chunk = await sync_to_async(next_slice)(sync_generator)
            ...
```

... so we (synchronously) fetch chunks of row before mapping them out asynchronously.

**Review by carltongibson — APPROVED** [review]
Looks nice @felixxm — thank you. 🏅

**Comment by felixxm** [comment]
buildbot, test on oracle.

**Review by charettes — APPROVED** [review]
Looking great!

Putting `RawQuerySet` specificities aside, I am really looking forward to see how well this perform when we perform queries of the form

```python
Author.objects.prefetch_related('collaborators__books')
```

I'm starting to wonder if a better implementation of `_async_fetch_all` would simply be

```python
async def _async_fetch_all(self):
    if self._result_cache is None:
        sync_to_async(self._fetch_all)()
```

when the backend doesn't support native async to limit the amount of sync-async hops required to iterate over non trivial data sets. Right now using anything that relies on `_async_fetch_all` (e.g. `__aiter__()`) will require `Ceil(N / 100) + P` `sync_to_async` hoop where `N` is the number of rows and `P={0,1}` depending on the usage of `prefetch_related` for no benefit since `_async_fetch_all` must completely entirely before the result set is available. This is an increase over what `sync_to_async(orm_operation)` use to do.

**Comment by felixxm** [comment]
Thanks both for reviews :1st_place_medal: 

> ... I am really looking forward to see how well this perform. I'm starting to wonder if a better implementation of `_async_fetch_all` would simply be
> 
> ```python
> async def _async_fetch_all(self):
>     if self._result_cache is None:
>         sync_to_async(self._fetch_all)()
> ```
> 
> when the backend doesn't support native async to limit the amount of sync-async hops required to iterate over non trivial data sets.
> 
> Right now using anything that relies on `_async_fetch_all` (e.g. `__aiter__()`) will require `Ceil(N / 100) + P` `sync_to_async` hoop where `N` is the number of rows and `P={0,1}` depending on the usage of `prefetch_related` for no benefit since `_async_fetch_all` must complete before the result set is available for iteration.
> 
> This is an increase over what `sync_to_async(orm_operation)` use to require.

Sounds like an noble idea :crown: , would you like to send PR with this cleanup?

**Comment by charettes** [comment]
> Sounds like an noble idea 👑 , would you like to send PR with this cleanup?

Done in https://github.com/django/django/pull/15658

**Pull request merged**